gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/GfxdSystemProcedures.java

```diff
@@ -1584,11 +1593,61 @@ public static void CREATE_OR_DROP_RESERVOIR_REGION(String reservoirRegionName,
     }
   }
 
+  /**
+   * Create or drop reservoir region for sampler.
+   *
+   * @param reservoirRegionName name of the reservoir region
+   * @param resolvedBaseName base table name with schema
+   * @param isDrop flag to indicate that the stored procedure is being invoked to drop the
+   *               reservoir region
+   * @param removeSampler flag to indicate that the cached sampler should also be removed
+   *                      along with the region. This flag will be a ignored if the `isDrop`
+   *                      flag is false.
+   */
+  public static void CREATE_OR_DROP_RESERVOIR_REGION_V2(String reservoirRegionName,
+      String resolvedBaseName, Boolean isDrop, Boolean removeSampler) throws SQLException {
+    try {
+      // check for permission on the sample table schema
+      LanguageConnectionContext lcc = ConnectionUtil.getCurrentLCC();
+      String currentUser = ((GenericLanguageConnectionContext)lcc).getUserName();
+      int dotIndex = resolvedBaseName.indexOf('.');
+      if (dotIndex == -1) {
+        throw new UnsupportedOperationException(
+            "Cannot created reservoir region for base name = " +
+                resolvedBaseName + " having no schema");
+      }
+      String schema = resolvedBaseName.substring(0, dotIndex);
+      CallbackFactoryProvider.getStoreCallbacks().checkSchemaPermission(
+          schema, currentUser);
+
+      // first create/drop locally
+      if (createOrDropReservoirRegion(reservoirRegionName, resolvedBaseName, isDrop,
+          removeSampler)) {
+        // don't send to other nodes or persist if local operation is unsuccessful
+        final Object[] args = new Object[] { reservoirRegionName,
+            resolvedBaseName, isDrop, removeSampler };
+        // send to other nodes
+        publishMessage(args, false,
+            GfxdSystemProcedureMessage.SysProcMethod.createOrDropReservoirRegionV2,
+            true, false);
+      }
+    } catch (StandardException se) {
+      throw PublicAPI.wrapStandardException(se);
+    } catch (Throwable t) {
+      throw TransactionResourceImpl.wrapInSQLException(t);
+    }
+  }
+
```

Comment:
Why code duplication? CREATE_OR_DROP_RESERVOIR_REGION should invoke CREATE_OR_DROP_RESERVOIR_REGION_V2 with last argument as true. In fact don't quite understand when would that be passed as false? Looks like that argument should always be true in which case it is not required at all, so the original procedure signature can remain as is without backward compatibility concerns.

Reply:
@sumwale, the `removeSampler` argument is passed false from here: `org.apache.spark.sql.sampling.ColumnFormatSamplingRelation#destroy`. This method does clean up of cached sampler separately and we wanted to keep this code untouched.
We can refactor this code and keep using `CREATE_OR_DROP_RESERVOIR_REGION` with remove sampler code executing always if this refactoring looks safe. OR we can keep this code as it is, in that case, the finally block of `destroy` method will be redundant.
Let me know whether this refactoring seems safe. I will rerun precheckin on the changes.

The code duplication was done to use different messages for both the stored procedures as I thought that would be needed for backward compatibility. Maybe my understanding was incorrect.

Reply:
> This method does clean up of cached sampler separately and we wanted to keep this code untouched.

Why?? Isn't it redundant and simply passing the "removeSampler" flag as true should be all that is required. Besides it will also test that new code a bit more.

> OR we can keep this code as it is, in that case, the finally block of destroy method will be redundant.

Yes, simply make this flag as always true and remove the finally block there. In other words remove the CREATE_OR_DROP_RESERVOIR_REGION_V2 and just have the original three argument CREATE_OR_DROP_RESERVOIR_REGION which internally will always do removeSampler. This way nothing related to backward compatibility is required and the code changes will be much simplified.

Reply:
@sumwale I have committed the suggested changes and will trigger precheckin soon.

While doing changes I noticed that by removing the finally block of `destroy` method, the remove sampler flow will be called from inside the procedure which will be called only when `isReservoirAsRegion`(which is set by `spark.sql.aqp.reservoirAsRegion`) flag is true. Earlier in the finally block the remove sampler flow was executed irrespective of this property. Does this seem appropriate?

Reply:
Ok, got it. The sampler removal block is required for non-region case too, so the finally block should be retained under if (!isReservoirAsRegion) condition. Afaik the API based sampling call on a DataFrame does not using a region rather just the in-memory route where isReservoirRegion is false.